### ISSUE TITLE
New SongSelectScene UI

### ISSUE THREAD

**Issue by achimmihca** [UltraStar-Deluxe/Play#265]
### Problem or question to be solved
I want to migrate the SongSelectScene to UIToolkit.

This is a rather complex scene and an important one because the player visits often, spents quite some time here, and makes several important configurations.
The UI needs to be self-explanatory, yet feature rich to find the desired songs (browse, filter, sort) and make needed changes (players, mics, voices, playlists, song queue, medleys, possibly more).

All features should be usable with only a controller (no pointer).

----
In the following "playlist" includes concepts for "song queue" and "medley queue"

SongSelectScene features:
- browse songs
    - jump to song
        - next/previous song
        - character quick jump bar
        - fuzzy search
        - jump to song by index
            - special search bar syntax (using hash: `#132`)
            - click on song index
    - filter songs
        - search bar
        - playlist
    - order songs (e.g., by artist, by title, by year, ...)

- select players
    - enable, disable player
    - assign mic to player (ATM this is done automatically (first come, first served))
    - assign voice to player (for duet song)

- manage playlists
    - create, remove, rename playlist
    - select, deselect playlist
    - add, remove, reorder songs
    - "use playlist for singing" vs. "use playlist for filtering songs"
        - creating playlist and using playlists are independent?

- online multiplayer
    - multiple remote players want to decide for a song
        - option to synchronize song browsing?
        - each player can decide for a song independently, but can continue only after all decide for the same
            - show player choice somewhere near song
        - only show songs that both players have in their song lib?
            - mark other songs as disabled / not available?

- show info
    - song info (is duet, has video, title, artist, etc.)
    - highscore info (local, remote)
    - etc.

----
I took a look at similar games for inspiration:

SingStar (PS3)
<img width="50%" src="https://user-images.githubusercontent.com/14967033/143769965-5ec04900-3d5b-44e1-9014-fbb6c6636bc6.png">

SingStar Celebration (PS4 I think)
Select song:
<img width="50%" src="https://user-images.githubusercontent.com/14967033/143769996-61456f1a-5ee8-4305-a8e0-2ef62531cd0f.png">

Adjust settings (still can change song with L2 / R2 buttons):
<img width="50%" src="https://user-images.githubusercontent.com/14967033/143770010-d6e9462d-09f0-41ab-ae93-b7bcef78e646.png">

UltraStar Deluxe
<img width="50%" src="https://user-images.githubusercontent.com/14967033/143770095-faf0d3e3-504e-4082-a828-c43f173be02e.jpg">

UltraStar World Party
<img width="50%" src="https://user-images.githubusercontent.com/14967033/143770217-bd01ff41-7bbb-491a-a9be-13cf940e4e0a.png">

Vocaluxe
<img width="50%" src="https://user-images.githubusercontent.com/14967033/143770133-c6c81b96-3f52-44eb-a507-e2938d8e7b9e.png">

----
### Your opinion on the topic

- SingStar Celebration shows a nice bar with letters to jump

- It's good to show the controller buttons for each action.
- Icons for L/R buttons are placed near their purpose (select next / previous item).
    - This info element could be a button so you can also click it.
        - Without controller, it could only show "<" and ">" without the corresponding controller button.

- The foreground / background contrast in SingStar Celebration has been increased compared to original SingStar
    - The contrast of Vocaluxe is terrible. You can barely tell foreground and background apart.
    - Thus, I argue for a simple background
- I think SingStar Celebration makes less use of icons.
    - An icon only conveys information if the user has seen it before, or sees it frequently, or it has been explained. If Icon and text are always shown together, it is redundant.
    - Thus, I argue that we should make use of icons only if they can **replace** text.
        - I like how UltraStar World Party has a legend of all icons in the popup menu.

- I like how UltraStar World Party always shows the search bar. 
    - However, I am unsure whether the search bar should be positioned at the top or bottom. At the top is more common I think.

- I like the large, separate popup (or slide in) menu bar at the right side in UltraStar World Party
    - Here is enough space to show several options and icon legend.
    - This could be scrollable for small screen sizes.
    - This could be opened via "hamburger button" (at the top right).
        - Uses little space when not needed and provides enough space when needed
- The current playlist could be shown as a second lane (or second grid) below the main song list (main song grid).
    - Changing playlist/main-list is matter of switching up/down between control areas.
    - Playlist will be modified by companion app.
        - Changes from the app should be visible immediately.
        - Thus, I don't think that the design of SingStar Celebration is a good fit for USPlay (seems to be hidden in a sub-menu). Vocaluxe approach seems better fit (always visible).
    - Space for this second lane is used by other controls when there is no playlist (everything is dynamic as needed)
- Interesting how Vocaluxe shows a menu over the song image / song video.
    - Nice reuse of screen space when the song has been decided.
    - Might be an option for something that is directly related to the song and needs only little space

**Comment by achimmihca** [comment]
I now have some ideas how this could be done. Will try some stuff soon.
Do you have further suggestions?

**Comment by achimmihca** [comment]
> The contrast of Vocaluxe is terrible. You can barely tell foreground and background apart.

The current contrast in USPlay is not good either. Currently, the color gradient is in the center. So far, this was not of interest.
But I think this distracts and makes stuff appear blurry where the most player attention should be.

Colors and background in UltraStar World party are really nice. They did a great job 👍 🚀.

**Comment by achimmihca** [comment]
I forgot the party mode modifiers (#243). Maybe this could be a sub-menu of the right hand side popup menu.

**Comment by bohning** [comment]
Wow, great overview and reasoning, thanks a lot. I think appearance/design goes a long way, so it makes sense to put in some time into these decisions.

- I like the idea of a "fuzzy search" (as e.g. Youtube offers it) if you don’t know the exact title (or don’t know exactly how artist/title are spelled), I could even imagine an additional option to search lyrics (if someone looks for a song, but doesn’t know either artist or title, just a fragment of the lyrics)
- Regarding song info: in addition to artist, title and year, this could also include song length (might be a decisive factor for/against a song, e.g. if it is very long), genre, edition and even creator.
- Song sorting by tabs/categories (not sure if had this also in mind with "order songs" above)
- Multi-level song search/filter would be an awesome feature ("I want to sing a Rock song from the 80s! What do you have?"), as outlined here: https://github.com/UltraStar-Deluxe/USDX/issues/568

**Comment by achimmihca** [comment]
Thanks for the feedback! These are all good points.

> "fuzzy search" (as e.g. Youtube offers it) 

You mean you search "my cool sng" and it suggests "my cool song" (with o)?

- There are some fuzzy search algorithm implemented in C# with MIT license on GitHub: https://github.com/search?q=fuzzy+string+language%3AC%23+license%3Amit+language%3AC%23&type=Repositories&ref=advsearch&l=C%23&l=C%23
- This one is Apache 2 License and seems to cover the basic algorithms: https://github.com/tylerjensen/FuzzyStrings

Can keep this in mind, but I think it is mostly independent of UI. Can be implemented afterwards.

> option to search lyrics

I agree it would be a nice feature, but it surely will be slower than other searches.

This requires smart loading of songs because the lyrics are not pre-loaded.
The search must be done on a separate Thread.

> genre, edition and even creator

I agree all song properties should be searchable.
Maybe one can combine the search bar with a dropdown for the searched properties (each as checkbox).

When a search has no results then the searched properties could be extended automatically (i.e., more checkboxes become checked). My code editor does this and it is great because everything is searched as needed and it is transparent to me what the searched includes.

> Regarding song info: [...] include song length [...], genre, edition and even creator.

I suggest to show song length always. This is also done on YouTube, Spotify etc.
The other properties are rather special and not available for all songs. I would show them on demand only (e.g. in a popup or overlay).

> Song sorting by tabs/categories

I think a tabbed interface (like in browsers) is too complex and out of scope.

By "order songs" I means order by artist, or by title, or by other properties, most popular, best rated, etc.

The filter handles what is in the search result.
The order defines what is presented first.

> Multi-level song search/filter would be an awesome feature

I doubt that the average player needs this.
Such complex queries are always a lot of UI work and most of the time the UI still sucks.
Thus, I would rather have an extended search syntax for the intrigued user, where you can have AND, OR, NOT: `genre="Rock" AND (language="German" OR language="Deutsch") AND year>=1980 AND year < 1990`

When using only the most basic stuff then I think one could build a general search for song properties.
One must be careful because this stuff gets complex quickly and people tend to always want more ;)
You quickly end up implementing a full SQL query including ORDER BY and other operators. But that's not the goal.
Anyway, this deserves an issue on its own.

**Comment by bohning** [comment]
By "tabs" I did not mean a tabbed interface as in a browser, but the option in USDX to set "Tabs = On" in settings. This lead to a two-level song selection. E.g. if sorting==language and tabs==on, then on the first level, all languages occuring in the available song set are the first level that can be browsed, then when selecting a certain language, only the songs of that language are shown.

Settings:
![image](https://user-images.githubusercontent.com/1497707/149641726-3cc31eeb-666b-4f58-9056-baee3cadf408.png)

Song Selection:
![image](https://user-images.githubusercontent.com/1497707/149641861-6fd9aa72-4270-4d87-8aa7-4bfc29a4d60d.png)

This was one way of handling large song sets, when someone only wants to sing German songs (and browse only those). I found this quite useful until I extended the search to support the language field, so maybe this tab feature is obsolete with suitable filter/search options.

**Comment by achimmihca** [comment]
> so maybe this tab feature is obsolete with suitable filter/search options.

I think so.

> The current playlist could be shown as a second lane (or second grid) below the main song list (main song grid).

After implementing the scene, I think a second lane (or grid) would complicate things too much.
Also, on mobile devices there is hardly enough space.

The main task here is to select (find / decide for) a song. A playlist can help with this.
But I felt that a second lane with songs was over the top for the supplementary task of "managing playlists".
Instead, there are now simple buttons to add/remove current song to playlist. And buttons to create/delete playlist. Done.

Maybe one could show a plain (vertical) UI list with the entries in the edit-playlist-popup. In this UI list one could reorder songs.
Or managing playlist could get a dedicated section in the settings.
Anyway, future work and not needed right now.
